src/interpreter/frame.rs

```diff
@@ -0,0 +1,923 @@
+
+                    Downcast(_, variant) => {
+                        use rustc::ty::layout::Layout::*;
+                        match *base_layout {
+                            General { discr, .. } => {
+                                return Ok(Lvalue {
+                                    ptr: base.ptr.offset(discr.size().bytes() as isize),
+                                    extra: LvalueExtra::DowncastVariant(variant),
+                                });
+                            }
+                            RawNullablePointer { .. } | StructWrappedNullablePointer { .. } => {
+                                return Ok(base);
+                            }
+                            _ => panic!("variant downcast on non-aggregate: {:?}", base_layout),
+                        }
+                    },
+
+                    Deref => {
+                        let pointee_ty = pointee_type(base_ty).expect("Deref of non-pointer");
+                        let ptr = self.gecx.memory.read_ptr(base.ptr)?;
+                        let extra = match pointee_ty.sty {
+                            ty::TySlice(_) | ty::TyStr => {
+                                let len_ptr = base.ptr.offset(self.gecx.memory.pointer_size as isize);
+                                let len = self.gecx.memory.read_usize(len_ptr)?;
+                                LvalueExtra::Length(len)
+                            }
+                            ty::TyTrait(_) => unimplemented!(),
+                            _ => LvalueExtra::None,
+                        };
+                        return Ok(Lvalue { ptr: ptr, extra: extra });
+                    }
+
+                    Index(ref operand) => {
+                        let elem_size = match base_ty.sty {
+                            ty::TyArray(elem_ty, _) |
+                            ty::TySlice(elem_ty) => self.type_size(elem_ty),
+                            _ => panic!("indexing expected an array or slice, got {:?}", base_ty),
+                        };
+                        let n_ptr = self.eval_operand(operand)?;
+                        let n = self.gecx.memory.read_usize(n_ptr)?;
+                        base.ptr.offset(n as isize * elem_size as isize)
+                    }
+
+                    ConstantIndex { .. } => unimplemented!(),
+                }
+            }
+        };
+
+        Ok(Lvalue { ptr: ptr, extra: LvalueExtra::None })
+    }
+
+    fn lvalue_ty(&self, lvalue: &mir::Lvalue<'tcx>) -> Ty<'tcx> {
+        self.monomorphize(self.frame.mir.lvalue_ty(self.gecx.tcx, lvalue).to_ty(self.gecx.tcx))
+    }
+
+    fn monomorphize(&self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        self.gecx.monomorphize(ty, self.frame.substs)
+    }
+
+    pub fn type_size(&self, ty: Ty<'tcx>) -> usize {
+        self.gecx.type_size(ty, self.frame.substs)
+    }
+
+    pub fn type_layout(&self, ty: Ty<'tcx>) -> &'tcx Layout {
+        self.gecx.type_layout(ty, self.frame.substs)
+    }
```

Comment:
Forwarding functions like the above three are why I made `FnEvalContext` deref to the `GlobalEvalContext` (like how `TyCtxt` works in rustc).